From 71e31dcce305e2190ac8cc3661a319a9ef9f173e Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Wed, 14 Oct 2020 18:17:23 +0800
Subject: [PATCH] PONRTSYS-8066:PHY: Adding specific mmd read and write
 functions to access MMD PHY registers

PHYLIB uses clause22 to access to clause 45 through indirection
registers, accesing EEE MMD registers to get EEE features unexpected
results.This will add specific read and write to access clause 45
registers.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/phy/lantiq.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/net/phy/lantiq.c b/drivers/net/phy/lantiq.c
index 5ea749123d74..19dcb01dc653 100644
--- a/drivers/net/phy/lantiq.c
+++ b/drivers/net/phy/lantiq.c
@@ -86,6 +86,22 @@ static __maybe_unused int vr9_gphy_mmd_write(struct phy_device *phydev,
 	return 0;
 }
 
+static int
+p31g_read_mmd_phyreg(struct phy_device *phydev, int regnum,
+		     int devnum, int ptrad)
+{
+	return mdiobus_read(phydev->mdio.bus, phydev->mdio.addr,
+				   MII_ADDR_C45 | (devnum << 16) | (regnum & 0xffff));
+}
+
+static void
+p31g_write_mmd_phyreg(struct phy_device *phydev, int regnum,
+		      int devnum, int ptrad, u32 val)
+{
+	regnum = MII_ADDR_C45 | ((devnum & 0x1f) << 16) | (regnum & 0xffff);
+	mdiobus_write(phydev->mdio.bus, phydev->mdio.addr, regnum, val);
+}
+
 static int vr9_gphy_config_init(struct phy_device *phydev)
 {
 	int err;
@@ -371,6 +387,8 @@ static struct phy_driver lantiq_phy[] = {
 		.read_status	= p31g_gphy_read_status,
 		.ack_interrupt	= vr9_gphy_ack_interrupt,
 		.config_intr	= vr9_gphy_config_intr,
+		.read_mmd_indirect = p31g_read_mmd_phyreg,
+		.write_mmd_indirect = p31g_write_mmd_phyreg,
 	},
 };
 
